pruners: Use overloads to catch improper configurations as type errors

ConnorBaker opened the pull request:
<!-- Thank you for creating a pull request! In general, we merge your pull request after it gets two or more approvals. To proceed to the review process by the maintainers, please make sure that the PR meets the following conditions: (1) it passes all CI checks, and (2) it is neither in the draft nor WIP state. If you wish to discuss the PR in the draft state or need any other help, please mention the Optuna development team in the PR. -->

## Motivation

Currently, improper arguments given to the pruners are only caught at runtime. Given the addition of `@overload` and `Literal` to `typing` and `typing_extensions`, we can provide warning about some improper uses of the pruners through the type checker.

## Description of the changes

This change adds a few overloads for the HyperBand, Successive Halving, and Threshold pruners to catch some invalid configurations at type checking time.

As an example:

```python
h1 = HyperbandPruner(1, "auto", 3, 0) # Works
# Argument of type Literal['cats'] cannot be assigned to parameter max_resource of type 
# int in function __init__
#   Literal['cats'] is incompatible with int
h2 = HyperbandPruner(1, "cats", 3, 0) # Type error! Not int or "auto".
# Argument of type Literal['cats'] cannot be assigned to parameter max_resource of type 
# int in function __init__
#   Literal['auto'] is incompatible with int
h3 = HyperbandPruner(1, "auto", 3, 1) # Type error! bootstrap_count>0 means we no auto
h4 = HyperbandPruner(max_resource="auto") # Works
h5 = HyperbandPruner(max_resource=1) # Works
```

toshihikoyanase reviewed (commented):
Thank you for your PR. I'm afraid we cannot apply the changes that require a new version of `typing_extension` since some ML libraries including `tensorflow` need older versions of `typing_extension`. Could you check the discussion in #3980 and #3990, please?

ConnorBaker commented:
It seems like the older version of `typing_extensions` supports `Literal` so I changed it back.

Although I wonder if I need to guard the import with `if TYPE_CHECKING` because I moved it out of the main requirements?

contramundum53 commented on optuna/pruners/_hyperband.py:
```diff
@@ -136,10 +141,33 @@ def objective(trial):
             See the details for :class:`~optuna.pruners.SuccessiveHalvingPruner`.
     """
 
+    @overload
+    def __init__(
+        self,
+        min_resource: int = 1,
+        max_resource: "Literal['auto']" = "auto",
+        reduction_factor: int = 3,
+        bootstrap_count: "Literal[0]" = 0,
+    ) -> None:
+        ...
+
+    @overload
+    def __init__(
+        self,
+        min_resource: int = 1,
+        # NOTE: We must provide a defualt value or else we cannot write the arguments in order.
+        # This value is not used; if the case a default is not provided the original init is called
+        # and max_resource is set to "auto".
+        max_resource: int = 100,
+        reduction_factor: int = 3,
+        bootstrap_count: int = 0,
+    ) -> None:
+        ...
```
I'm sorry I'm not sure about the behavior of the type checker, but do we need these overloads? In principle the type `Union[Literal['auto'], int]` should be enough, isn't it?

contramundum53 commented on tests/pruners_tests/test_successive_halving.py:
```diff
@@ -253,10 +256,15 @@ def test_successive_halving_pruner_min_early_stopping_rate_parameter() -> None:
 
     with pytest.raises(ValueError):
-        optuna.pruners.SuccessiveHalvingPruner(bootstrap_count=-1)
+        # Because the default min_resource is "auto", the bootstrap_count must be zero.
+        # The type checker correctly detects this type error.
+        optuna.pruners.SuccessiveHalvingPruner(bootstrap_count=-1)  # type: ignore
```
Maybe after we are able to detect these errors by type-checking, we could remove these tests according to the test policy.

contramundum53 replied:
@ConnorBaker We concluded that we can remove these invalid tests in this PR if we can detect them by type-checking. How about that?

contramundum53 commented:
I think we can take over.

ConnorBaker commented:
Apologies, I haven't had much time as of late. Please do!
